### PR TITLE
Multiple outputs

### PR DESCRIPTION
**Pull request by abadams**
This is a branch that adds the ability to compile pipelines that produce multiple output Funcs of different sizes and dimensionalities - e.g. an image pyramid, or a yuv420 frame. The type that represents an imaging pipeline is now "Pipeline", though Funcs can still be used to represent single-output pipelines (they contain an internal Pipeline object that they forward calls to).

So you can now return either a Func or a Pipeline (which can represent one or more outputs) from your Generator::build() method. This required making build() not a virtual function, so I had to remove the override keyword in all the test generators. Apart from that existing user code should be unchanged.

Most of the new logic in Pipeline.cpp is a rewrite (and clean up) of the logic that was in Func.cpp.

See the test/generators/pyramid_generator.cpp example and also test/correctness/multiple_outputs.cpp for some examples. More suggestions for testing are welcome.

### CODE REVIEW

File: test/generator/pyramid_generator.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#include "Halide.h"
+
+namespace {
+
+class Pyramid : public Halide::Generator<Pyramid> {
+public:
+    ImageParam input{ Float(32), 2, "input" };
```
Comment (steven-johnson) [inline test/generator/pyramid_generator.cpp]: To be uber-pedantic, you could add a GeneratorParam to specify the number of levels

**Comment by steven-johnson** [comment]
This is gonna take a while to review :-/

**Comment by abadams** [comment]
I'm out on vacation next week, so there's no rush.

**Comment by abadams** [comment]
Ok, back from vacation. I made the number of pyramid levels a generator param, and deleted some dead code. PTAL.

File: src/Pipeline.cpp [hunk 2]
```diff
@@ -0,0 +1,1057 @@
...
+            return true;
+        } else if (other.arg.is_buffer() && !arg.is_buffer()) {
+            return false;
+        } else {
+            return arg.name < other.arg.name;
+        }
+    }
+};
+
+struct PipelineContents {
```
Comment (dsharletg) [inline src/Pipeline.cpp]: Is it possible to make this an opaque struct, and move the definition/usages to Pipeline.cpp?

Comment (dsharletg) [inline src/Pipeline.cpp]: Wait... I'm an idiot! It's already Pipeline.cpp

**Comment by dsharletg** [comment]
LGTM, though I'd really like to find a way to support use cases where an output is only store_root, and not just compute_root (not for this PR). The biggest draw for this to me was the possibility of improving the locality of pyramid construction, but if all the outputs need to be compute root, that doesn't happen.

**Comment by abadams** [comment]
Merged in master again. Anyone else want to take a look at this?

File: test/generator/pyramid_aottest.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    Image<float> input(1024, 1024);
+
+    // Put some junk in the input. Keep it to small integers so the float averaging stays exact.
+    for (int y = 0; y < input.height(); y++) {
+        for (int x = 0; x < input.width(); x++) {
+            input(x, y) = ((x * 17 + y)/8) % 32;
+        }
+    }
+
+    vector<Image<float> > levels(10);
```
Comment (aam) [inline test/generator/pyramid_aottest.cpp]: You don't have to put space between angle brackets.

File: src/StorageFlattening.cpp [hunk 4]
```diff
@@ -256,12 +260,16 @@ class FlattenDimensions : public IRMutator {
         vector<ProvideValue> values;
         flatten_provide_values(values, provide);
 
+        bool is_output = false;
+        for (Function f : outputs) {
+            is_output |= f.name() == provide->name;
```
Comment (aam) [inline src/StorageFlattening.cpp]: Do  you need to iterate through all outputs or can you break after finding the match?

**Pull request merged**